### PR TITLE
Move discovery to DNS names

### LINKED ISSUE #2830: Use Pod DNS names instead of IP addresses for discovery and TLS certificates
Related to https://github.com/elastic/cloud-on-k8s/issues/2823.

We configure nodes to advertise and discover themselves through their IP addresses.
For that reason, we are likely to regenerate TLS certificates when a Pod is restarted since its IP address changed.

It does not help with [this bug](https://github.com/elastic/cloud-on-k8s/issues/2823) where Elasticsearch may serve a certificate with the wrong IP address associated.

Things would be simpler if we only manipulated Pod DNS names instead of IP addresses, and make more sense for TLS certificates which are usually bound to a DNS name, not an IP address.

Questions (to investigate):
- can DNS caching be an issue for nodes to discover themselves?
- are there cases (security constraints) where Pods are not allowed/not supposed to rely on DNS?
- any issue with the DNS server will lead to Elasticsearch not behaving correctly: are we ok with that?
- what's the correct DNS name to use? `<pod-name>`, `<pod-name>.<namespace>`

### PR DESCRIPTION
**Pull request by pebrc**
Fixes #2830
Related #3654 

This PR changes Elasticsearch discovery to use DNS hostnames instead of IP addresses. To replicate the discussion on #3654 : we think the potential negative impact of DNS caching is manageable and outweighs the advantage of simplifying discovery.

We could in theory remove the IP address from the ECK-generated transport certificates now and save a few certificate regenerations when a node is scheduled on another Pod. I left it in place because I believe the frequency of these regenerations if fairly low but open to reconsider.

### CODE REVIEW

File: pkg/controller/elasticsearch/sset/pod.go [hunk 1]
```diff
@@ -20,6 +20,18 @@ import (
...
+// HeadlessServiceName returns the name of the headless service for the given StatefulSet.
+func HeadlessServiceName(ssetName string) string {
+	// just use the sset name
+	return ssetName
+}
+
+//PodDNSName returns the DNS resolvable name of the given pod resolvable within the namespace.
+func PodDNSName(pod corev1.Pod) string {
+	ssetName := pod.Labels[label.StatefulSetNameLabelName]
+	return fmt.Sprintf("%s.%s", pod.Name, HeadlessServiceName(ssetName))
```
Comment (sebgl) [inline pkg/controller/elasticsearch/sset/pod.go:32]: I'm wondering if we should consider how a user may override the Pod hostname and subdomain fields: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-hostname-and-subdomain-fields
But I think that's really a corner case that is probably safe to ignore for now. It probably has other impacts.

**Review by sebgl — APPROVED** [review]
LGTM

Side note: using DNS names only in the certificates (https://github.com/elastic/cloud-on-k8s/issues/2830) would help fixing https://github.com/elastic/cloud-on-k8s/issues/2832 and https://github.com/elastic/cloud-on-k8s/issues/2833.

File: pkg/controller/elasticsearch/sset/pod.go [hunk 2]
```diff
@@ -20,6 +20,18 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/utils/stringsutil"
 )
 
+// HeadlessServiceName returns the name of the headless service for the given StatefulSet.
+func HeadlessServiceName(ssetName string) string {
+	// just use the sset name
+	return ssetName
+}
+
+//PodDNSName returns the DNS resolvable name of the given pod resolvable within the namespace.
```
Comment (thbkrkr) [inline pkg/controller/elasticsearch/sset/pod.go:29]:
```suggestion
// PodDNSName returns the DNS resolvable name of the given pod resolvable within the namespace.
```

**Comment by pebrc** [comment]
Closing in the light of https://github.com/elastic/cloud-on-k8s/issues/3723#issuecomment-690249624

**Pull request closed without merging**